src/classes/RD2_ValidationService.cls

```diff
@@ -35,46 +35,114 @@
     */
-    public void validate(List<npe03__Recurring_Donation__c> rds) {
+    private List<npe03__Recurring_Donation__c> oldRds = new List<npe03__Recurring_Donation__c>();
 
-        Set<Id> accountIds = new Set<Id>();
-        Set<Id> contactIds = new Set<Id>();
+    /**
+    * @description Accounts on the Rds
+    */
+    private Map<Id,Account> accountById {
+        get {
+            if (accountById == null) {
+                accountById = getAccounts();
+            }
+            return accountById;
+        } set;
+    }
 
-        for (npe03__Recurring_Donation__c rd : rds) {
-            if (rd.npe03__Organization__c != null) {
-                accountIds.add(rd.npe03__Organization__c);
-                contactIds.add(rd.npe03__Contact__c);
+    /***
+    * @descripton Ids of Recurring Donations that can have donor modified 
+    */
+    private Set<Id> adjustableRdIds {
+        get {
+            if (adjustableRdIds == null) {
+                adjustableRdIds = getDonorAdjustableRDIds();
             }
+            return adjustableRdIds;
         }
-        accountIds.remove(null);
-        contactIds.remove(null);
+        set;
+    }
 
-        Map<Id,Account> accountsById = new Map<Id,Account>([
-            SELECT Id, npe01__SYSTEMIsIndividual__c,
-            (
-                SELECT Id
-                FROM Contacts
-                WHERE Id IN :contactIds
-            )
-            FROM Account
-            WHERE Id IN :accountIds
-            ]);
+    /**
+    * @description Constructor for the RD2_ValidationService class
+    * @param rds New rds that will be validated
+    */
+    public RD2_ValidationService(List<npe03__Recurring_Donation__c> rds) {
+        this.rds = rds;
+    }
+
+    /**
+    * @description Configure recurring donation trigger old list to the service class
+    * @param oldRds ld Recurring Donation records available in the update/delete dml operation.
+    * @return RD2_ValidationService
+    */
+    public RD2_ValidationService withOldRecurringDonations(List<npe03__Recurring_Donation__c> oldRds) {
+        this.oldRds = oldRds;
+        return this;
+    }
 
+    /**
+    * @description Validate RD data on insert
+    */
+    public void validate() {
         for (npe03__Recurring_Donation__c rd : rds) {
             validateActiveRecurringDonation(rd);
-            validateDonor(rd, accountsById.get(rd.npe03__Organization__c));
+            validateDonor(rd, accountById.get(rd.npe03__Organization__c));
+        }
+    }
+
+    /**
+    * @description Validate RD data on update
+    */
+    public void validateOnChange() {
+         for (Integer i = 0; i < rds.size(); i++) {
+            npe03__Recurring_Donation__c rd = rds[i];
+            npe03__Recurring_Donation__c oldRd = oldRds[i];
+
+            validateActiveRecurringDonation(rd);
+            
+            Boolean donorChangeApprove = validateDonorChange(rd, oldRd);
```

Comment:
Boolean var have name as a question, so here we can have a generic validity check var: `Boolean isValid = validateDonorChange(rd, oldRd);`